app/telemetry/handler.py

```diff
@@ -71,3 +81,44 @@ def get_path(path_id):
     for k, v in TELEMETRY_PATH_IDS.items():
         if v == path_id:
             return k
+
+
+def record_metrics(shim, submission_timestamp):
+    """If METRICS_SAMPLE_RATE is set and greater than 0, then log metrics for the event.
+        The metrics are logged in a structured event to enable google cloud log-based metrics to aggregate them
+
+        metrics sampled:
+            time from adserver generating the impression to now
+            time from telemetry submission to now
+    """
+    try:
+        # range of 0-1000, to sample in increments of 1/1000
+        metrics_sample_rate = os.environ.get("METRICS_SAMPLE_RATE", "0")
+
+        if not metrics_sample_rate.isdecimal():
+            return
+
+        if int(metrics_sample_rate) <= 0 or random.randrange(0, 1000) >= int(metrics_sample_rate):
+            return
+
+        log_client = google.cloud.logging.Client()
+        log_client.setup_logging()
+
+        submission_timestamp_millis = int(datetime.fromisoformat(submission_timestamp).timestamp() * 1000)
+
+        _, encoded_data, _ = shim.split(",")
+        padded_data = encoded_data + '=' * (-len(encoded_data)%4)
+        kevel_json = json.loads(base64.b64decode(padded_data))
+
+        kevel_timestamp = kevel_json['ts']
+
+        now_millis = int(time.time() * 1000)
+
+        glean_latency_millis = now_millis - submission_timestamp_millis
+        adserver_latency_millis = now_millis - kevel_timestamp
+
+        logging.info("metrics", extra={"json_fields": {"glean_latency": glean_latency_millis, "adserver_latency": adserver_latency_millis}})
```

Comment:
Should we also capture the device and Firefox version?  It would highlight any differences between the ping types.